File: src/Shared/SIMF.Common/SaudiRegions.cs
namespace SIMF.Common;

/// <summary>
/// The 13 official Saudi administrative regions, used for the
/// birth-location dropdown (FR #4): a Saudi registrant picks a region; everyone
/// else types their place of birth free-form "as in passport". A fixed constant
/// (the regions are official and stable), mirrored by the app's
/// <c>saudi_regions.dart</c> — there is no shared runtime between .NET and
/// Flutter. The selected region's localized <b>name</b> is stored in the existing
/// free-text <c>UserProfile.PlaceOfBirth</c> column, so this adds no schema and
/// no new validation — it only constrains the input for Saudis.
/// </summary>
public static class SaudiRegions
{
    /// <summary>One region: a stable <paramref name="Code"/> plus its Arabic and
    /// English display names.</summary>
    public sealed record Region(string Code, string Arabic, string English);

    /// <summary>The 13 regions, in the conventional order.</summary>
    public static readonly IReadOnlyList<Region> All =
    [
        new("riyadh", "منطقة الرياض", "Riyadh"),
        new("makkah", "منطقة مكة المكرمة", "Makkah"),
        new("madinah", "منطقة المدينة المنورة", "Al Madinah"),
        new("eastern", "المنطقة الشرقية", "Eastern Province"),
        new("asir", "منطقة عسير", "Asir"),
        new("tabuk", "منطقة تبوك", "Tabuk"),
        new("hail", "منطقة حائل", "Hail"),
        new("northern", "منطقة الحدود الشمالية", "Northern Borders"),
        new("jazan", "منطقة جازان", "Jazan"),
        new("najran", "منطقة نجران", "Najran"),
        new("bahah", "منطقة الباحة", "Al Bahah"),
        new("jawf", "منطقة الجوف", "Al Jawf"),
        new("qassim", "منطقة القصيم", "Al Qassim"),
    ];

    /// <summary>The region whose <see cref="Region.Code"/> equals <paramref name="code"/>, or null.</summary>
    public static Region? ByCode(string? code) =>
        All.FirstOrDefault(r => r.Code == code);

    /// <summary>
    /// The region whose Arabic or English name equals <paramref name="name"/>
    /// (trimmed), or null — maps a stored place-of-birth string (saved in either
    /// language) back to a region so the dropdown survives a UI-culture switch.
    /// </summary>
    public static Region? ByName(string? name)
    {
        var n = name?.Trim() ?? string.Empty;
        return n.Length == 0
            ? null
            : All.FirstOrDefault(r => r.Arabic == n || r.English == n);
    }
}

File: src/Mobile/simf_app/lib/features/account/saudi_regions.dart
/// D-469 — the 13 official Saudi administrative regions, for the birth-location
/// field: a Saudi registrant picks a region from a dropdown; everyone else
/// types their place of birth free-form "as in passport". Mirrors the server
/// `SaudiRegions` (no shared runtime). The selected region's localized name is
/// stored in the existing free-text `placeOfBirth` field — no schema change.
library;

/// One region: a stable [code] plus its [arabic] / [english] display names.
class SaudiRegion {
  const SaudiRegion(this.code, this.arabic, this.english);
  final String code;
  final String arabic;
  final String english;

  /// The display name for the current locale.
  String name({required bool isArabic}) => isArabic ? arabic : english;
}

/// The 13 regions, in the conventional order.
const List<SaudiRegion> saudiRegions = <SaudiRegion>[
  SaudiRegion('riyadh', 'منطقة الرياض', 'Riyadh'),
  SaudiRegion('makkah', 'منطقة مكة المكرمة', 'Makkah'),
  SaudiRegion('madinah', 'منطقة المدينة المنورة', 'Al Madinah'),
  SaudiRegion('eastern', 'المنطقة الشرقية', 'Eastern Province'),
  SaudiRegion('asir', 'منطقة عسير', 'Asir'),
  SaudiRegion('tabuk', 'منطقة تبوك', 'Tabuk'),
  SaudiRegion('hail', 'منطقة حائل', 'Hail'),
  SaudiRegion('northern', 'منطقة الحدود الشمالية', 'Northern Borders'),
  SaudiRegion('jazan', 'منطقة جازان', 'Jazan'),
  SaudiRegion('najran', 'منطقة نجران', 'Najran'),
  SaudiRegion('bahah', 'منطقة الباحة', 'Al Bahah'),
  SaudiRegion('jawf', 'منطقة الجوف', 'Al Jawf'),
  SaudiRegion('qassim', 'منطقة القصيم', 'Al Qassim'),
];

/// The region whose Arabic or English name equals [name] (trimmed), or null —
/// used to map a stored place-of-birth string back to a dropdown selection.
SaudiRegion? regionByName(String? name) {
  final n = name?.trim() ?? '';
  if (n.isEmpty) {
    return null;
  }
  for (final r in saudiRegions) {
    if (r.arabic == n || r.english == n) {
      return r;
    }
  }
  return null;
}
